reana/reana_benchmark/monitor.py

```diff
@@ -0,0 +1,151 @@
+# This file is part of REANA.
+# Copyright (C) 2021 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Responsible for monitoring K8s cluster, DB connections."""
+
+import json
+import time
+from pathlib import Path
+from typing import Dict, Any, List
+from collections import defaultdict
+import subprocess
+from abc import abstractmethod, ABC
+
+from reana.reana_benchmark.utils import get_utc_now_timestamp, logger
+
+
+class BaseMetric(ABC):
+    """Base class for other metrics."""
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """Name of the metric."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def _collect(self) -> Any:
+        raise NotImplementedError
+
+    def collect(self) -> Dict[str, Any]:  # noqa: D102
+        result = self._collect()
+        return {
+            self.name: result,
+        }
+
+
+class NumberOfDBConnectionsMetric(BaseMetric):
+    """Count number of server processes in REANA DB."""
+
+    name = "db_connections_number"
+
+    def _collect(self) -> Any:
+        cmd = [
+            "kubectl",
+            "exec",
+            "deployment/reana-db",
+            "--",
+            "psql",
+            "-U",
+            "reana",
+            "-c",
+            "SELECT COUNT(*) FROM pg_stat_activity;",
+        ]
+        output = subprocess.check_output(cmd).decode("ascii")
+        result = int(output.splitlines()[2].strip())
+        return result
+
+
+class WorkflowPodsMetric(BaseMetric):
+    """Count number of job and batch jobs in different phases."""
+
+    name = "workflows_pods_status"
+
+    @staticmethod
+    def _filter(pods: List[Dict], name_contains: str) -> List[Dict]:
+        selected_pods = []
+        for pod in pods:
+            name = pod.get("metadata", {}).get("name", "")
+            if name_contains in name:
+                selected_pods.append(pod)
+        return selected_pods
+
+    @staticmethod
+    def _count(pods: List[Dict]) -> Dict[str, int]:
+        statistics = defaultdict(lambda: 0)
+        for pod in pods:
+            phase = pod.get("status", {}).get("phase")
```

Comment:
*question*: The `phase` value is not representing exactly what `kubectl get pods` (`STATUS` column) is displaying. It has `Pending`, `Running` and `Succeed` statuses. In addition, it has `Failed` status which, for some reason, appears for batch pods at the end when workflows are *successfully* finished. 

I guess this happens because batch pods enter `NotReady` state because the workflow engine successfully finished and exits. You cannot easily get `NotReady` state from kubectl JSON output, you need to deduct it from container statuses.

What do you think about the situation? Any ideas on how to get a more accurate status?

Reply:
The `phase` values are described [here](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase). I think it's a pretty accurate metric for us to use. But this case with successfully finished `Failed` status could be confusing.. I guess one way how we could solve it is to include the `containerStatuses` to the final output (downsize is that it will make it much more verbose..). Here is an example:
```
kubectl get pod -o=jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.status.phase}{"\t"}{.status.containerStatuses[*].state}{"\n"}{end}'
...
reana-server-69c494847b-fxldq   Running {"running":{"startedAt":"2021-11-23T08:09:29Z"}} {"running":{"startedAt":"2021-11-23T08:09:29Z"}}
...
```
Then it would be visible that the `Failed` status happens because the containers terminated. But not sure if we need it, could be confusing to have too much output..

Reply:
Example of a JSON output about pod `status` for a single pod from `kubectl get pods -o json` command:

```json
"status": {
      "conditions": [
          {
              "lastProbeTime": null,
              "lastTransitionTime": "2021-11-11T16:15:07Z",
              "status": "True",
              "type": "Initialized"
          },
          {
              "lastProbeTime": null,
              "lastTransitionTime": "2021-11-11T16:15:12Z",
              "status": "True",
              "type": "Ready"
          },
          {
              "lastProbeTime": null,
              "lastTransitionTime": "2021-11-11T16:15:12Z",
              "status": "True",
              "type": "ContainersReady"
          },
          {
              "lastProbeTime": null,
              "lastTransitionTime": "2021-11-11T16:15:07Z",
              "status": "True",
              "type": "PodScheduled"
          }
      ],
      "containerStatuses": [
          {
              "containerID": "containerd://78d3916f2dfca74427f8b0cc38cc344a746a2594c8fd6f746133e7a10b0cfdbb",
              "image": "docker.io/reanahub/reana-workflow-controller:latest",
              "imageID": "sha256:75e1ea2066193703f73e1c71fce92969a7b9c47b8af3e307dc2d5ee812e2d40d",
              "lastState": {},
              "name": "job-status-consumer",
              "ready": true,
              "restartCount": 0,
              "started": true,
              "state": {
                  "running": {
                      "startedAt": "2021-11-11T16:15:10Z"
                  }
              }
          },
          {
              "containerID": "containerd://4b915673af6c64cbec168cc8512969507544365d8ba35b3900f836836d384a26",
              "image": "docker.io/reanahub/reana-workflow-controller:latest",
              "imageID": "sha256:75e1ea2066193703f73e1c71fce92969a7b9c47b8af3e307dc2d5ee812e2d40d",
              "lastState": {},
              "name": "rest-api",
              "ready": true,
              "restartCount": 0,
              "started": true,
              "state": {
                  "running": {
                      "startedAt": "2021-11-11T16:15:11Z"
                  }
              }
          }
      ],
      "hostIP": "172.18.0.2",
      "phase": "Running",
      "podIP": "10.244.0.117",
      "podIPs": [
          {
              "ip": "10.244.0.117"
          }
      ],
      "qosClass": "BestEffort",
      "startTime": "2021-11-11T16:15:07Z"
}
```

As described above `phase` can show pods as `Failed` while they are terminating and it is not so straighforward how to extract more precise info from the above JSON. Some checks can be done on `containerStatuses` or, maybe, in `conditions`.

Reply:
The current goal for monitoring command is to collect data that later can be visualized along with execution progress plot. With additional `containerStatuses` output it is not clear how to visualize it (there is a `phase` but also one or more containers in `containerStatuses`). I would prefer to reduce it to one field (`Running: 5`, `Failed: 1`, etc.) but the question is how :)

Reply:
If I understood correctly when the pod is terminated `SIGKILL` is sent by k8s and apparently `job controller` takes too much time to exit so status is `137` and marked as `Failed`.

Example of JSON:

```json
"containerStatuses": [
      {
        "containerID": "containerd://400831aac7443eba9a41228abf264d1adfa63d8279b3a4b71f34f819af7df420",
        "image": "docker.io/reanahub/reana-job-controller:latest",
        "imageID": "sha256:268ad55769e69f1c41d448fb7b54a95dc074cba5e2e1cfa6a2642d8484756551",
        "lastState": {},
        "name": "job-controller",
        "ready": false,
        "restartCount": 0,
        "started": false,
        "state": {
          "terminated": {
            "containerID": "containerd://400831aac7443eba9a41228abf264d1adfa63d8279b3a4b71f34f819af7df420",
            "exitCode": 137,
            "finishedAt": "2021-11-24T10:28:32Z",
            "reason": "Error",
            "startedAt": "2021-11-24T10:27:44Z"
          }
        }
      },
      {
        "containerID": "containerd://d5e4a82bfe1acd86411058ebaa0302dd8a6d9632b1a07d6f849789071918b1b3",
        "image": "docker.io/reanahub/reana-workflow-engine-serial:latest",
        "imageID": "sha256:5756537cf4d2673d2725e0824f6f6883062c363229d6c81a500b19a1487b8641",
        "lastState": {},
        "name": "workflow-engine",
        "ready": false,
        "restartCount": 0,
        "started": false,
        "state": {
          "terminated": {
            "containerID": "containerd://d5e4a82bfe1acd86411058ebaa0302dd8a6d9632b1a07d6f849789071918b1b3",
            "exitCode": 0,
            "finishedAt": "2021-11-24T10:28:02Z",
            "reason": "Completed",
            "startedAt": "2021-11-24T10:27:44Z"
          }
        }
      }
    ],
```

Check timestamps in `containerStatuses`. `workflow-engine` finished at `10:28:02` so around that time REANA asks k8s to delete pod. Take a look at `job-controller` finish time - `10:28:32`. Exactly 30 seconds after `workflow-engine`, 30 seconds is a **default time** in k8s for graceful shutdown of a pod ([details](https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-terminating-with-grace)).

The same happened in other batch pods - `job controller` is not terminating within 30 seconds when k8s asks for it. @tiborsimko

*suggestion:* I think the above case deserves a separate issue. I can leave the `phase` value in the metric for this PR. I assume when we will figure out why the exit code is `137` instead of `0`, the `phase` value will show `Succeeded` status for batch pods.

Reply:
Opened the issue to fix `Failed` phase for batch pods. We can proceed forward with this PR and leave `phase` as it is.